### DISCUSSION TITLE
Tools & CLIs that are useful for projects.

### DISCUSSION THREAD

**Discussion by Raynos** [voxpelli/types-in-js#3] [Best Practices / Patterns]
There are various tools that are useful for working with jsdoc.

Here's an overview of some that I know of, and it would be great if this list can be expanded upon

 - `typescript` / `tsc` ; you need this.
 - `type-coverage` ; The `type-coverage` command helps you identify if your project is 50%, 90% or 99% statically typed and it can let you know which JS statements require extra JSDoc annotations.
 - `eslint` ; The eslint project along with [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint) has useful rules for JSDoc projects.

There are some more opinionated tools that can streamline some steps

 - `tsdocstandard` ; This rolls up a bunch of eslint plugins / rules and helps you with JSDoc annotations

**Comment by voxpelli** [comment]
### Test / linting

* [`tsd`](https://github.com/SamVerschueren/tsd) is useful if one wants to add verification of ones types, eg. ensuring that the public facing types doesn't break by becoming narrower or different.

### Useful conversion tools

* [`json-schema-to-typescript`](https://github.com/bcherny/json-schema-to-typescript) – enables one to convert a JSON schema to a `.d.ts` which one can then import into ones JSDoc doing eg. `/** @type {import('/filename.d.ts').Foo} */` or `/** @typedef {import('/filename.d.ts').Foo} Foo */`. This is useful as JSON Schema can validate ones code at runtime and thus ensure that eg. user input actually complies with the assumptions of ones JSDoc types (a JSON Schema can also be even stricter than checking just types, taking care of all ones validation needs, which eg. [fastify](https://www.fastify.io/) is making good use of in their routes)

**Comment by voxpelli** [comment]
Another really good tool:

* [`typesync`](https://github.com/jeffijoe/typesync) – I run `typesync --dry` all the time to figure out if I'm missing any `@types/` for my dependencies. Good way to find out if there are more type definitions out there that can enhance ones code.

And when it comes to linting:

* [`eslint-plugin-jsdoc`](https://github.com/gajus/eslint-plugin-jsdoc) is really handy, either with a custom config or with my reusable one [`@voxpelli/eslint-config-jsdoc-ts`](https://github.com/voxpelli/eslint-config-jsdoc-ts)

**Comment by niksy** [comment]
What is the status of linting JSDoc comments with `typescript-eslint` or `eslint-plugin-jsdoc`?

I see there is [open discussion](https://github.com/typescript-eslint/typescript-eslint/issues/2939) regarding supporting more features inside comments.

Currently, I have personal config with `eslint-plugin-jsdoc` features enabled/disabled for TypeScript, and rely on `tsc` to warn about any potential errors, but it would be nice to have same rules from `typescript-eslint` available.

**Reply by voxpelli** [reply]
I do the same as you @niksy with my [`@voxpelli/eslint-config-jsdoc-ts`](https://github.com/voxpelli/eslint-config-jsdoc-ts).

At times I play with the idea of looking more into `typescript-eslint`, but bringing an additional parser into it all has so far made me feel that it adds more complexity then it helps. But I should probably revisit that soon.

**Reply by Raynos** [reply]
I put as many rules as I can into https://github.com/Raynos/tsdocstandard you can check that out.
